include/envoy/network/dns.h

```diff
@@ -39,7 +39,8 @@ class DnsResolver {
    * @return if non-null, a handle that can be used to cancel the resolution.
    *         This is only valid until the invocation of callback or ~DnsResolver().
    */
-  virtual ActiveDnsQuery* resolve(const std::string& dns_name, ResolveCb callback) PURE;
+  virtual ActiveDnsQuery* resolve(const std::string& dns_name, uint32_t port,
```

Comment:
I think there are a few things here:
1) The Address interface having a port is somewhat orthogonal IMO to this issue. Right now we spit out Address from DNS that effectively have no port, and then we make a new address with the port merged in.
2) I do agree that this interface and how one might use/document it is confusing. I guess I could be convinced that passing 0 as the DNS address port in the config engages SRV support, but IMO that is too much magic and we will get a lot of questions about it.

I can go either way on having port passed into the DNS interfaces. On one hand, I agree that conceptually it's strange because A records don't have ports. On the other hand, I think it does make the calling code simpler, especially when you start adding in SRV support.

I guess my vote (a weak vote) would be to spit port out of the DNS interfaces only in the case of SRV, but not engage SRV magically via passing a 0 port. I would add a new function, resolveSrv() which does not take port as an argument (but addresses that come out will have a port). Then, calling code, based on config, can either call resolve() or resolveSrv(). In the case of standard resolve() they can merge the port in after the fact like they do now. Otherwise they get a port in the address that comes out. (The alternative would be to have a new interface for Address that does not have port, but then we need a whole new set of callbacks for SRV resolving that spits out the ones with ports).

The main question IMO is how to actually engage SRV in the config and then document it. I think there are a few options here:
1) Actually do use port 0 and document it. E.g., "tcp://foo.service:0"
2) Have some extra config element like "dns_use_srv: true" which overrides port
3) Do something special in the "URL" parsing like "tcp://foo.service:srv"

I actually kind of like 3, but I'm open to suggestions.

Anyway, I can't say that I have a very strong opinion here. I think we could go a bunch of different ways.